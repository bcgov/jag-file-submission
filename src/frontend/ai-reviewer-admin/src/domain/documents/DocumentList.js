import React, { useState } from "react";
import PropTypes from "prop-types";
import { VscJson } from "react-icons/vsc";
import { MdDeleteForever, MdEdit, MdLibraryAdd } from "react-icons/md";

import "./DocumentList.scss";

export default function DocumentList({ configurations, onDelete }) {  
  return (
    <div className="document-type-list">
      <div className="table-header">
        <span className="d-none d-sm-inline col-sm-3">Document Type</span>
        <span className="d-none d-sm-inline col-sm-3">Document Description</span>
        <span className="d-none d-sm-inline col-sm-3">Project ID</span>
        <span className="d-none d-sm-inline col-sm-3">Action</span>
      </div>
      <ul>
        {configurations.map((configuration) => (
          <li key={configuration.id}>
            <span className="col-12 d-inline d-sm-none label"><VscJson size="24" color="#FCBA19" strokeWidth="1" />{configuration.id}</span>

            <span className="col-6 d-inline d-sm-none">Document Type</span>
            <span className="col-6 col-sm-3 label">{configuration.documentType.type}</span>
            <span className="col-6 d-inline d-sm-none">Document Description</span>
            <span className="col-6 col-sm-3 label">{configuration.documentType.description}</span>
            <span className="col-6 d-inline d-sm-none">Project ID</span>
            <span className="col-6 col-sm-3 label">{configuration.projectId}</span>
            <span className="col-12 col-sm-3 pull-right">
              <MdEdit size="24" color="#FCBA19" />
              <span
                  role="button"
                  tabIndex={-1}
                  onClick={(e) => onDelete(configuration.id)}
                  data-testid={"delete-"+configuration.id}
                >
                <MdDeleteForever size="24" className="delete-icon"/>
              </span>
            </span>
          </li>
        ))}
        <li><span className="col"><MdLibraryAdd size="24" color="#FCBA19" /></span></li>
      </ul>
    </div>
  );
}

DocumentList.propTypes = {
  configurations: PropTypes.arrayOf(PropTypes.object),
  onDelete: PropTypes.func.isRequired,
};

DocumentList.defaultProps = {
  configurations: [],
};
