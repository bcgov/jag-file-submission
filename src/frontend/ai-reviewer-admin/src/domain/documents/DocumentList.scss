.document-type-list {
  .table-header {
    display: flex;
    flex-wrap: wrap;
    font-weight: bold;
  }
  ul {
    margin: 5px 0px;
    padding: 0px;
  }
  li {
    display: flex;
    flex-wrap: wrap;
    list-style-type: none;
    border: 1px solid #bfbfbf;
    padding: 10px 0px;
    margin: 10px 0px;
  }
  .label {
    font-weight: bold;
    margin-bottom: 0px;
  }
  svg {
    margin-right: 10px;
    vertical-align: top;
  }
  @media screen and (max-width: 575px) {
    .pull-right {
      text-align: right;
    }
  }
}
