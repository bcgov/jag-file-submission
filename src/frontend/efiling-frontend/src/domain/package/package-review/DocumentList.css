.header {
    display: flex;
    flex-wrap: wrap;
    font-weight: bold;
}
.label {
    font-weight: bold;
    margin-bottom: 0px;
}
ul {
    margin: 5px 0px;
    padding: 0px;
}
li {
    display: flex;
    flex-wrap: wrap;
    list-style-type: none;
    border: 2px solid #E1BEBD;
    padding: 10px 0px;
}