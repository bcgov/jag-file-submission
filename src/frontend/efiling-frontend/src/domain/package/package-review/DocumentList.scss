.ct-document-list {
  .header {
    display: flex;
    flex-wrap: wrap;
    font-weight: bold;
  }
  .label {
    font-weight: bold;
    margin-bottom: 0px;
  }
  ul {
    margin: 5px 0px;
    padding: 0px;
  }
  li {
    display: flex;
    flex-wrap: wrap;
    list-style-type: none;
    border: 1px solid #bfbfbf;
    padding: 10px 0px;
    margin: 10px 0px;
  }
  .REJ, .CCOR {
    border: 1px solid #e88387 !important;
  }

  li span {
    text-overflow: ellipsis;
    overflow: hidden;
  }
  li span.file-cell {
    padding-top: 1px;
  }
  .hidden {
    display: none;
  }
  .modal-body p,
  .modal-body .row div {
    text-overflow: ellipsis;
    overflow: hidden;
  }
  .modal-body .data {
    font-weight: bold;
  }
  .modal-footer {
    display: block;
    align-items: normal;
    justify-content: normal;
  }
  .pull-left {
    float: left;
  }
  .pull-right {
    float: right;
  }
  svg {
    margin-right: 10px;
  }
}
