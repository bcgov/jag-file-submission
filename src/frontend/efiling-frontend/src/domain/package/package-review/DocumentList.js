/* eslint-disable react/jsx-one-expression-per-line */
import React, { useState } from "react";
import PropTypes from "prop-types";
import { MdHelp, MdDescription } from "react-icons/md";
import moment from "moment";

import Modal from "react-bootstrap/Modal";
import { Button } from "shared-components";
import { errorRedirect } from "../../../modules/helpers/errorRedirect";
import { withdrawTooltipData } from "../../../modules/test-data/withdrawTooltipData";
import { isClick, isEnter } from "../../../modules/helpers/eventUtil";
import {
  downloadSubmittedDocument,
  withdrawSubmittedDocument,
} from "./PackageReviewService";
import "./DocumentList.scss";
import BcGovTooltip from "./tooltip/Tooltip";

export default function DocumentList({
  packageId,
  documents,
  reloadDocumentList,
}) {
  const [showModal, setShowModal] = useState(false);
  const [modalData, setModalData] = useState({
    description: "",
    documentProperties: {
      name: "",
    },
    filingDate: "",
    status: "",
  });

  const handleModalClose = () => setShowModal(false);
  const handleModalConfirm = () => {
    setShowModal(false);

    withdrawSubmittedDocument(packageId, modalData)
      .then(() => reloadDocumentList())
      .catch((err) => errorRedirect(sessionStorage.getItem("errorUrl"), err));
  };

  function handleDownloadFileEvent(e, document) {
    if (isClick(e) || isEnter(e)) {
      downloadSubmittedDocument(packageId, document).catch((err) =>
        errorRedirect(sessionStorage.getItem("errorUrl"), err)
      );
    }
  }

  function handleWithdrawFileEvent(e, document) {
    if (isClick(e) || isEnter(e)) {
      setModalData(document);
      setShowModal(true);
    }
  }

  return (
    <div className="ct-document-list">
      <div className="header">
        <span className="d-none d-lg-inline col-lg-4">File Name</span>
        <span className="d-none d-lg-inline col-lg-3">Document Type</span>
        <span className="d-none d-lg-inline col-lg-3">Status</span>
        <span className="d-none d-lg-inline col-lg-2">Action (s)</span>
      </div>
      <ul>
        {documents &&
          documents.map((document) => (
            <li key={document.identifier}>
              <span className="label col-sm-4 d-lg-none">File Name:</span>
              <span className="col-sm-8 col-lg-4 file-cell">
                <MdDescription size={30} color="#FCBA19" />
                <span
                  className="file-href"
                  role="button"
                  data-testid="btn-download-document"
                  tabIndex={0}
                  onClick={(e) => handleDownloadFileEvent(e, document)}
                  onKeyDown={(e) => handleDownloadFileEvent(e, document)}
                >
                  {document.documentProperties.name}
                </span>
              </span>
              <span className="label col-sm-4 d-lg-none">Document Type:</span>
              <span className="col-sm-8 col-lg-3">{document.description}</span>
              <span className="label col-sm-4 d-lg-none">Status:</span>
              <span className="col-sm-8 col-lg-3">
                {document.status.description}
              </span>
              <span className="label col-sm-4 d-lg-none">Action (s):</span>
              <span className="col-sm-8 col-lg-2 file-cell">
                <span
                  id={`withdraw_${document.identifier}`}
                  className="file-href"
                  role="button"
                  tabIndex={0}
                  onClick={(e) => handleWithdrawFileEvent(e, document)}
                  onKeyDown={(e) => handleWithdrawFileEvent(e, document)}
                >
                  withdraw
                </span>
                <MdHelp
                  style={{
                    fontSize: "20px",
                    marginLeft: "10px",
                    color: "#7598ca",
                  }}
                  id="withdraw-tooltip"
                />
                <BcGovTooltip
                  target={`withdraw-tooltip`}
                  content={withdrawTooltipData}
                />
              </span>
            </li>
          ))}
      </ul>

      <Modal
        show={showModal}
        onHide={handleModalClose}
        className="ct-document-list"
      >
        <Modal.Header closeButton>
          <Modal.Title>Withdraw Document</Modal.Title>
        </Modal.Header>

        <Modal.Body>
          <p>
            Please confirm the withdrawl of document{" "}
            {modalData.documentProperties.name} from your latest submission.
            This will remove the document permanently from your submitted
            package number
            {packageId}.
          </p>

          <div className="row">
            <div className="col-sm-4">Package Number:</div>
            <div className="col-sm-8 data">{packageId}</div>
          </div>
          <div className="row">
            <div className="col-sm-4">Document Status:</div>
            <div className="col-sm-8 data">{modalData.status.description}</div>
          </div>
          <div className="row">
            <div className="col-sm-4">Date Filed:</div>
            <div className="col-sm-8 data">
              {moment(modalData.filingDate).format("DD-MMM-YYYY")}
            </div>
          </div>
          <div className="row">
            <div className="col-sm-4">Document Type:</div>
            <div className="col-sm-8 data">{modalData.description}</div>
          </div>
          <div className="row">
            <div className="col-sm-4">File Name:</div>
            <div className="col-sm-8 data">
              {modalData.documentProperties.name}
            </div>
          </div>
        </Modal.Body>
        <Modal.Footer>
          <Button
            onClick={handleModalClose}
            label="Cancel"
            styling="btn btn-secondary pull-left"
          />
          <Button
            onClick={handleModalConfirm}
            label="Confirm"
            styling="btn btn-primary pull-right"
          />
        </Modal.Footer>
      </Modal>
    </div>
  );
}

DocumentList.propTypes = {
  packageId: PropTypes.number.isRequired,
  documents: PropTypes.arrayOf(
    PropTypes.shape({
      identifier: PropTypes.string.isRequired,
      documentProperties: PropTypes.shape({
        type: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
      }),
      status: PropTypes.shape({
        description: PropTypes.string.isRequired,
      }),
    })
  ),
  reloadDocumentList: PropTypes.func.isRequired,
};

DocumentList.defaultProps = {
  documents: [],
};
