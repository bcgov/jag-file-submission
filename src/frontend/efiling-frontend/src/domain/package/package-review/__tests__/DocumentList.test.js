import React from "react";
import { render, waitFor, fireEvent, getByText } from "@testing-library/react";
import DocumentList from "../DocumentList";

const pkgRvwService = require("../PackageReviewService");

describe("DocumentList Component Testsuite", () => {
  const packageId = 1;
  const documents = [
    {
      identifier: "1",
      filingDate: "2020-05-05T00:00:00.000Z",
      documentProperties: {
        type: "AFF",
        name: "test-document.pdf",
      },
      status: {
        description: "Submitted",
      },
    },
  ];
  const reloadDocumentList = jest.fn();

  beforeEach(() => {
    window.open = jest.fn();
  });

  test("Matches the snapshot", async () => {
    const { asFragment } = render(
      <DocumentList
        packageId={packageId}
        documents={documents}
        reloadDocumentList={reloadDocumentList}
      />
    );
    await waitFor(() => {});
    expect(asFragment()).toMatchSnapshot();
  });

  test("Download document (click) success", async () => {
    // stub out service to return valid response.
    pkgRvwService.downloadSubmittedDocument = jest.fn(() => Promise.resolve());

    const { container } = render(
      <DocumentList
        packageId={packageId}
        documents={documents}
        reloadDocumentList={reloadDocumentList}
      />
    );
    await waitFor(() => {});

    // get the span wrapping the file link, click it.
    fireEvent.click(
      getByText(container, `${documents[0].documentProperties.name}`)
    );
    await waitFor(() => {});

    expect.assertions(1);
    expect(pkgRvwService.downloadSubmittedDocument()).resolves.not.toThrow();
  });

  test("Download document (click) fail", async () => {
    const errorMessage = "Network Error";
    // stub out service to return failed response.
    pkgRvwService.downloadSubmittedDocument = jest.fn(() =>
      Promise.reject(new Error(errorMessage))
    );

    const { container } = render(
      <DocumentList
        packageId={packageId}
        documents={documents}
        reloadDocumentList={reloadDocumentList}
      />
    );
    await waitFor(() => {});

    // get the span wrapping the file link, click it.
    fireEvent.click(
      getByText(container, `${documents[0].documentProperties.name}`)
    );
    await waitFor(() => {});

    expect.assertions(1);
    expect(pkgRvwService.downloadSubmittedDocument()).rejects.toThrow(
      errorMessage
    );
  });

  test("Download document (keyDown) success keyCode=13", async () => {
    // stub out service to return valid response.
    pkgRvwService.downloadSubmittedDocument = jest.fn(() => Promise.resolve());

    const { container } = render(
      <DocumentList
        packageId={packageId}
        documents={documents}
        reloadDocumentList={reloadDocumentList}
      />
    );
    await waitFor(() => {});

    // get the span wrapping the file link, click it.
    fireEvent.keyDown(
      getByText(container, `${documents[0].documentProperties.name}`),
      {
        key: "Enter",
        keyCode: "13",
      }
    );
    await waitFor(() => {});

    expect.assertions(1);
    expect(pkgRvwService.downloadSubmittedDocument()).resolves.not.toThrow();
  });

  test("Download document (keyDown) success keyCode=32", async () => {
    // stub out service to return valid response.
    pkgRvwService.downloadSubmittedDocument = jest.fn(() => Promise.resolve());

    const { container } = render(
      <DocumentList
        packageId={packageId}
        documents={documents}
        reloadDocumentList={reloadDocumentList}
      />
    );
    await waitFor(() => {});

    // get the span wrapping the file link, click it.
    fireEvent.keyDown(
      getByText(container, `${documents[0].documentProperties.name}`),
      {
        key: "Space",
        keyCode: "32",
      }
    );
    await waitFor(() => {});

    expect.assertions(1);
    expect(pkgRvwService.downloadSubmittedDocument).not.toHaveBeenCalled();
  });

  test("Download document (keyDown) fail", async () => {
    const errorMessage = "Network Error";
    // stub out service to return failed response.
    pkgRvwService.downloadSubmittedDocument = jest.fn(() =>
      Promise.reject(new Error(errorMessage))
    );

    const { container } = render(
      <DocumentList
        packageId={packageId}
        documents={documents}
        reloadDocumentList={reloadDocumentList}
      />
    );
    await waitFor(() => {});

    // get the span wrapping the file link, click it.
    fireEvent.keyDown(
      getByText(container, `${documents[0].documentProperties.name}`),
      {
        key: "Enter",
        keyCode: "13",
      }
    );
    await waitFor(() => {});

    expect.assertions(1);
    expect(pkgRvwService.downloadSubmittedDocument()).rejects.toThrow(
      errorMessage
    );
  });

  test("Withdraw document (click) success", async () => {
    // stub out service to return valid response.
    pkgRvwService.withdrawSubmittedDocument = jest.fn(() => Promise.resolve());

    const { container } = render(
      <DocumentList
        packageId={packageId}
        documents={documents}
        reloadDocumentList={reloadDocumentList}
      />
    );
    await waitFor(() => {});

    // get the span wrapping the file link, click it.
    fireEvent.click(getByText(container, "withdraw"));
    await waitFor(() => {});

    expect.assertions(1);
    expect(pkgRvwService.withdrawSubmittedDocument()).resolves.not.toThrow();
  });

  test("Withdraw document (click) fail", async () => {
    const errorMessage = "Network Error";
    // stub out service to return failed response.
    pkgRvwService.withdrawSubmittedDocument = jest.fn(() =>
      Promise.reject(new Error(errorMessage))
    );

    const { container } = render(
      <DocumentList
        packageId={packageId}
        documents={documents}
        reloadDocumentList={reloadDocumentList}
      />
    );
    await waitFor(() => {});

    // get the span wrapping the file link, click it.
    fireEvent.click(getByText(container, "withdraw"));
    await waitFor(() => {});

    expect.assertions(1);
    expect(pkgRvwService.withdrawSubmittedDocument()).rejects.toThrow(
      errorMessage
    );
  });

  test("Withdraw document (keyDown=13) success", async () => {
    // stub out service to return valid response.
    pkgRvwService.withdrawSubmittedDocument = jest.fn(() => Promise.resolve());

    const { container } = render(
      <DocumentList
        packageId={packageId}
        documents={documents}
        reloadDocumentList={reloadDocumentList}
      />
    );
    await waitFor(() => {});

    // get the span wrapping the file link, click it.
    fireEvent.keyDown(getByText(container, "withdraw"), {
      key: "Enter",
      keyCode: "13",
    });
    await waitFor(() => {});

    expect.assertions(1);
    expect(pkgRvwService.withdrawSubmittedDocument()).resolves.not.toThrow();

    // get the span wrapping the file link, click it.
    fireEvent.keyDown(getByText(container, "withdraw"));
    await waitFor(() => {});

    expect.assertions(2);
    expect(pkgRvwService.withdrawSubmittedDocument()).resolves.not.toThrow();
  });

  test("Withdraw document (keyDown) fail", async () => {
    const errorMessage = "Network Error";
    // stub out service to return failed response.
    pkgRvwService.withdrawSubmittedDocument = jest.fn(() =>
      Promise.reject(new Error(errorMessage))
    );

    const { container } = render(
      <DocumentList
        packageId={packageId}
        documents={documents}
        reloadDocumentList={reloadDocumentList}
      />
    );
    await waitFor(() => {});

    // get the span wrapping the file link, click it.
    fireEvent.keyDown(getByText(container, "withdraw"), {
      key: "Enter",
      keyCode: "13",
    });
    await waitFor(() => {});

    expect.assertions(1);
    expect(pkgRvwService.withdrawSubmittedDocument()).rejects.toThrow(
      errorMessage
    );
  });
});
