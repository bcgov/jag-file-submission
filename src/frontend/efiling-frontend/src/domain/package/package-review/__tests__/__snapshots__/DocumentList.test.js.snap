// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentList Component Testsuite Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ct-document-list"
  >
    <div
      class="header"
    >
      <span
        class="d-none d-lg-inline col-lg-4"
      >
        File Name
      </span>
      <span
        class="d-none d-lg-inline col-lg-3"
      >
        Document Type
      </span>
      <span
        class="d-none d-lg-inline col-lg-3"
      >
        Status
      </span>
      <span
        class="d-none d-lg-inline col-lg-2"
      >
        Action (s)
      </span>
    </div>
    <ul>
      <li>
        <span
          class="label col-sm-4 d-lg-none"
        >
          File Name:
        </span>
        <span
          class="col-sm-8 col-lg-4 file-cell"
        >
          <svg
            color="#FCBA19"
            fill="currentColor"
            height="30"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(252, 186, 25);"
            viewBox="0 0 24 24"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14 2H6c-1.1 0-1.99.9-1.99 2L4 20c0 1.1.89 2 1.99 2H18c1.1 0 2-.9 2-2V8l-6-6zm2 16H8v-2h8v2zm0-4H8v-2h8v2zm-3-5V3.5L18.5 9H13z"
            />
          </svg>
          <span
            class="file-href"
            data-testid="btn-download-document"
            role="button"
            tabindex="0"
          >
            test-document.pdf
          </span>
        </span>
        <span
          class="label col-sm-4 d-lg-none"
        >
          Document Type:
        </span>
        <span
          class="col-sm-8 col-lg-3"
        />
        <span
          class="label col-sm-4 d-lg-none"
        >
          Status:
        </span>
        <span
          class="col-sm-8 col-lg-3"
        >
          Submitted
        </span>
        <span
          class="label col-sm-4 d-lg-none"
        >
          Action (s):
        </span>
        <span
          class="col-sm-8 col-lg-2 file-cell"
        >
          <span
            class="file-href"
            id="withdraw_1"
            role="button"
            tabindex="0"
          >
            withdraw
          </span>
          <svg
            fill="currentColor"
            height="1em"
            id="withdraw-tooltip"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(117, 152, 202); font-size: 20px; margin-left: 10px;"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 17h-2v-2h2v2zm2.07-7.75l-.9.92C13.45 12.9 13 13.5 13 15h-2v-.5c0-1.1.45-2.1 1.17-2.83l1.24-1.26c.37-.36.59-.86.59-1.41 0-1.1-.9-2-2-2s-2 .9-2 2H8c0-2.21 1.79-4 4-4s4 1.79 4 4c0 .88-.36 1.68-.93 2.25z"
            />
          </svg>
          <div
            class="bcgov-withdraw-tooltip "
          />
        </span>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
